### PR TITLE
Python 3 support + CANbus backend

### PR DESCRIPTION
**Pull request by cowo78**
Developed these changes to support Python3 and to use CANbus as a transport backend. Not yet production ready, just have a look to see if you're interested. Once I have something reliable I'll add the PSOC C code as reference.

### CODE REVIEW

**Comment by Arachnid** [comment]
Thanks for the PR! Based on a quick skim, looks good. Do you think you could send the indentation changes as a separate PR, though? I assume you used a prettyprinter or your editor's support to automatically fix it up, so it should be fairly easy to do, and it would make reading this PR a lot easier.

**Comment by cowo78** [comment]
In the master bootload.py and cyacd_test.py are tab-indented, the others are space-indented. Just unified to space (personal taste). Want me to switch indentation back to the situation in master?

**Comment by Arachnid** [comment]
Space is definitely preferable. But the PR's very hard to review with a combination of indendation changes and code changes. Could you convert it back and submit the indentation fix as a separate PR?

**Comment by cowo78** [comment]
Restored indentation in bootload.py.
Found a bug in Cypress bootloader component v1.40 and v1.50 (confirmed by Cypress). Should be fixed in v1.60, added a note in README for future reference.

File: README.md [hunk 1]
```diff
@@ -36,3 +36,11 @@ and --newapp and --nonewapp, respectively.
 
 Cyflash is still quite new, and should be considered beta-quality software.
 Pull requests and bug reports are most welcome.
+
+
+Cypres Bootloader metadata component bug
```
Comment (Arachnid) [inline README.md]: +s

File: cyflash/bootload.py [hunk 2]
```diff
@@ -120,9 +168,13 @@ def __init__(self, session, out):
 		self.row_ranges = {}
 
 	def bootload(self, data, downgrade, newapp):
+		print ("Entering bootload")
```
Comment (Arachnid) [inline cyflash/bootload.py]: Were these for debugging? They seem excessively verbose for normal operation.

File: cyflash/bootload.py [hunk 3]
```diff
@@ -132,7 +184,9 @@ def bootload(self, data, downgrade, newapp):
 		self.session.exit_bootloader()
 
 	def verify_row_ranges(self, data):
-		for array_id, array in data.arrays.iteritems():
+		# Python2 only
```
Comment (Arachnid) [inline cyflash/bootload.py]: Could you remove these commented-out lines?

File: cyflash/bootload.py [hunk 4]
```diff
@@ -184,8 +241,11 @@ def check_metadata(self, data, downgrade, newapp):
 	def write_rows(self, data):
 		total = sum(len(x) for x in data.arrays.values())
 		i = 0
-		for array_id, array in data.arrays.iteritems():
-			for row_number, row in array.iteritems():
+		# Python 2
```
Comment (Arachnid) [inline cyflash/bootload.py]: As above, could you remove the commented lines?

File: cyflash/bootload.py [hunk 5]
```diff
@@ -194,6 +254,7 @@ def write_rows(self, data):
 						"Checksum does not match in array %d row %d. Expected %.2x, got %.2x! Aborting." % (
 							array_id, row_number, row.checksum, actual_checksum))
 				self.progress("Uploading data", i, total)
+				self.out.write(" - array {} row {}, {} data bytes".format(array_id, row_number, len(row.data)))
```
Comment (Arachnid) [inline cyflash/bootload.py]: Was this for debugging?

File: cyflash/cyacd.py [hunk 6]
```diff
@@ -12,14 +14,20 @@ def read(cls, data, line=None):
...
             raise ValueError("Row specified %d bytes of data, but got %d"
                              % (data_length, len(self.data)))
-        (checksum,) = struct.unpack('B', data[-1])
-        data_checksum = 0x100 - (sum(ord(x) for x in data[:-1]) & 0xFF)
+        # Python2. data is already a bytes object in Py3
+        #(checksum,) = struct.unpack('B', data[-1])
+        checksum = data[-1]
+        # Python2, see above
+        #data_checksum = 0x100 - (sum(ord(x) for x in data[:-1]) & 0xFF)
+        data_checksum = 0x100 - (sum(x for x in data[:-1]) & 0xFF)
```
Comment (Arachnid) [inline cyflash/cyacd.py]: The genexp is redundant now.

File: cyflash/cyacd.py [hunk 7]
```diff
@@ -12,14 +14,20 @@ def read(cls, data, line=None):
...
         self.array_id, self.row_number, data_length = struct.unpack('>BHH', data[:5])
         self.data = data[5:-1]
         if len(self.data) != data_length:
             raise ValueError("Row specified %d bytes of data, but got %d"
                              % (data_length, len(self.data)))
-        (checksum,) = struct.unpack('B', data[-1])
-        data_checksum = 0x100 - (sum(ord(x) for x in data[:-1]) & 0xFF)
+        # Python2. data is already a bytes object in Py3
+        #(checksum,) = struct.unpack('B', data[-1])
+        checksum = data[-1]
```
Comment (Arachnid) [inline cyflash/cyacd.py]: Does this still work in Python 2?

**Comment by Arachnid** [comment]
Thanks again for the contribution! The additions look good. Does this break compatibility with Python 2, however?

**Comment by cowo78** [comment]
See latest commits.

- Substituted print() with proper stream output. Matter of personal taste, I like more verbose output, feel free to remove
- Removed redundant list generation

The main issue is: yes, **this breaks compatibility with python2**

**Comment by Arachnid** [comment]
Ah, that's a bit of a dealbreaker; I'm not prepared to break this for everyone using Python 2. :/

Shouldn't most of these changes be fairly easy to make compatible with both versions? For instance, `binascii.unhexlify()` exists in both 2 and 3.

**Comment by cowo78** [comment]
Well, I'm not suggesting to break py2 compatibility, but you may want to branch a py3 version. I'm not sure I can spare the time to put in py2/py3 compatible code and to test it, that's the problem.

**Comment by cowo78** [comment]
I've seen that [this](https://pythonhosted.org/six/) package provides utilities to wrap differences between py2 and py3. What about bringing it in and having it do the hard stuff?

**Comment by Arachnid** [comment]
That seems like a reasonable, if slightly clunky solution. I'm not likely to have much time to work on this in the near future myself, however. :/

**Comment by Arachnid** [comment]
Oh, fantastic job, thanks so much! Have you determined if this works in Python 2?

**Comment by cowo78** [comment]
Yes, tested on 2.7 (CANbus transport).

**Comment by Arachnid** [comment]
Fantastic! Can you add six to the list of deps in setup.py? I've got a few other comments inline, too.

File: CANbus_Bootloader.c [hunk 8]
```diff
@@ -0,0 +1,273 @@
+/** \file
```
Comment (Arachnid) [inline CANbus_Bootloader.c]: If you want to include this, could you put this in a subdirectory, with a comment up top on how to use it?

File: cyflash/bootload.py [hunk 9]
```diff
@@ -206,6 +277,17 @@ def progress(self, message=None, current=None, total=None):
 
 def main():
 	args = parser.parse_args()
+
+	if (args.logging_config):
+		import sys
```
Comment (Arachnid) [inline cyflash/bootload.py]: What's "import sys" for, here? I don't believe "import logging" is necessary, either.

Comment (cowo78) [inline cyflash/bootload.py]: Removed sys. Logging needs to be initialized before python-can if a logging configuration file has been specified on the command line.

File: cyflash/bootload.py (hunk 5 above)
Comment (Arachnid) [inline cyflash/bootload.py]: Is this line still needed? It seems likely to clutter up the status output.

File: CANbus_Bootloader.c (hunk 8 above)
Comment (cowo78) [inline CANbus_Bootloader.c]: Done

**Comment by Arachnid** [comment]
Thanks so much for your persistence on this!

**Comment by cowo78** [comment]
Persistence = just too lazy to maintain a fork on my own.
Hope it will be useful to someone else, will post something on cypress forum.

**Pull request merged**